File: app/lib/data/bundle.dart
import 'dart:typed_data';

import '../models/relic.dart';

/// A relic with file attachments stores them as a single "bundle" blob — the
/// attachment bytes concatenated in manifest order. The manifest (each
/// [Attachment.size]) gives the offsets to slice them back apart. Keeping it to
/// one blob means the sync layer and the Worker treat it exactly like any other
/// single-blob relic (no schema/quota/cleanup changes).

/// Concatenate attachment payloads into one bundle, in order.
Uint8List packBundle(List<Uint8List> parts) {
  final total = parts.fold<int>(0, (n, p) => n + p.length);
  final out = Uint8List(total);
  var off = 0;
  for (final p in parts) {
    out.setRange(off, off + p.length, p);
    off += p.length;
  }
  return out;
}

/// Slice a single attachment's bytes out of [bundle] using the manifest order.
/// Returns null if the manifest doesn't line up with the bundle length.
Uint8List? sliceAttachment(
    Uint8List bundle, List<Attachment> manifest, String attId) {
  var off = 0;
  for (final a in manifest) {
    if (a.id == attId) {
      if (off + a.size > bundle.length) return null;
      return Uint8List.sublistView(bundle, off, off + a.size);
    }
    off += a.size;
  }
  return null;
}

File: relic-cli/src/bundle.rs
//! Bundle packing: multiple attachments concatenated into one blob, sliced back
//! apart using the manifest (`Attachment.size` gives each part's length).
//! Mirrors `app/lib/data/bundle.dart` so bundles interoperate with the app.

use crate::models::Attachment;

/// Concatenate attachment payloads in manifest order.
pub fn pack(parts: &[Vec<u8>]) -> Vec<u8> {
    let total: usize = parts.iter().map(Vec::len).sum();
    let mut out = Vec::with_capacity(total);
    for p in parts {
        out.extend_from_slice(p);
    }
    out
}

/// Slice one attachment's bytes out of `bundle` by manifest order. Returns
/// `None` if the id isn't in the manifest or the manifest overruns the bundle.
pub fn slice<'a>(bundle: &'a [u8], manifest: &[Attachment], id: &str) -> Option<&'a [u8]> {
    let mut off = 0usize;
    for a in manifest {
        let size = a.size as usize;
        if a.id == id {
            return if off + size <= bundle.len() { Some(&bundle[off..off + size]) } else { None };
        }
        off += size;
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    fn att(id: &str, size: u64) -> Attachment {
        Attachment { id: id.into(), name: format!("{id}.bin"), mime: None, size }
    }

    #[test]
    fn pack_then_slice_round_trips() {
        let parts = vec![b"AAA".to_vec(), b"BBBB".to_vec(), b"C".to_vec()];
        let manifest = vec![att("a", 3), att("b", 4), att("c", 1)];
        let bundle = pack(&parts);
        assert_eq!(bundle, b"AAABBBBC");
        assert_eq!(slice(&bundle, &manifest, "a"), Some(&b"AAA"[..]));
        assert_eq!(slice(&bundle, &manifest, "b"), Some(&b"BBBB"[..]));
        assert_eq!(slice(&bundle, &manifest, "c"), Some(&b"C"[..]));
        assert_eq!(slice(&bundle, &manifest, "missing"), None);
    }

    #[test]
    fn slice_rejects_manifest_overrun() {
        let manifest = vec![att("a", 3), att("b", 99)];
        assert_eq!(slice(b"AAAB", &manifest, "b"), None);
    }
}
